ports/nrf/common-hal/busio/UART.c

```diff
@@ -35,58 +35,329 @@
+            MICROPY_VM_HOOK_LOOP
+#endif
+        }
+
+        // copy received data
+        if ( self->rx_count > 0 ) {
+            memcpy(data, self->buffer, self->rx_count);
+            data += self->rx_count;
+            remain -= self->rx_count;
+
+            self->rx_count = 0;
+        }
+
+        // exit if complete or time up
+        if ( !remain || !(ticks_ms - start_ticks < self->timeout_ms) ) {
+            break;
+        }
+
+        // prepare next receiving
+        const size_t cnt = MIN(self->bufsize, remain);
+        self->rx_count = -1;
+        _VERIFY_ERR(nrfx_uarte_rx(&self->uarte, self->buffer, cnt));
+    }
+    
+    // queue 1-byte transfer for rx_characters_available()
+    if ( self->rx_count == 0 ) {
+        self->rx_count = -1;
+        _VERIFY_ERR(nrfx_uarte_rx(&self->uarte, self->buffer, 1));
+    }
+
+    return len - remain;
 }
 
 // Write characters.
 size_t common_hal_busio_uart_write(busio_uart_obj_t *self, const uint8_t *data, size_t len, int *errcode) {
-    mp_raise_NotImplementedError(translate("busio.UART not yet implemented"));
-    return 0;
+    if ( nrf_uarte_tx_pin_get(self->uarte.p_reg) == NRF_UARTE_PSEL_DISCONNECTED ) {
+        mp_raise_ValueError(translate("No TX pin"));
+    }
+
+    if ( len == 0 ) return 0;
+
+    uint64_t start_ticks = ticks_ms;
+
+    // Wait for on-going transfer to complete
+    while ( nrfx_uarte_tx_in_progress(&self->uarte) && (ticks_ms - start_ticks < self->timeout_ms) ) {
+#ifdef MICROPY_VM_HOOK_LOOP
+        MICROPY_VM_HOOK_LOOP
+#endif
+    }
+
+    // Time up
+    if ( !(ticks_ms - start_ticks < self->timeout_ms) ) {
+        *errcode = MP_EAGAIN;
+        return MP_STREAM_ERROR;
+    }
+
+    // EasyDMA can only access SRAM
+    uint8_t * tx_buf = (uint8_t*) data;
+    if ( !nrfx_is_in_ram(data) ) {
+        tx_buf = (uint8_t *) gc_alloc(len, false, false);
+        memcpy(tx_buf, data, len);
+    }
+
+    (*errcode) = nrfx_uarte_tx(&self->uarte, tx_buf, len);
+    _VERIFY_ERR(*errcode);
+    (*errcode) = 0;
+
+    while ( nrfx_uarte_tx_in_progress(&self->uarte) && (ticks_ms - start_ticks < self->timeout_ms) ) {
+#ifdef MICROPY_VM_HOOK_LOOP
+        MICROPY_VM_HOOK_LOOP
+#endif
+    }
+
+    if ( !nrfx_is_in_ram(data) ) {
+        gc_free(tx_buf);
+    }
+
+    return len;
 }
 
 uint32_t common_hal_busio_uart_get_baudrate(busio_uart_obj_t *self) {
-    mp_raise_NotImplementedError(translate("busio.UART not yet implemented"));
     return self->baudrate;
 }
 
 void common_hal_busio_uart_set_baudrate(busio_uart_obj_t *self, uint32_t baudrate) {
-    mp_raise_NotImplementedError(translate("busio.UART not yet implemented"));
     self->baudrate = baudrate;
+    nrf_uarte_baudrate_set(self->uarte.p_reg, get_nrf_baud(baudrate));
 }
 
 uint32_t common_hal_busio_uart_rx_characters_available(busio_uart_obj_t *self) {
+    return (self->rx_count > 0) ? self->rx_count : 0;
+}
+
+void common_hal_busio_uart_clear_rx_buffer(busio_uart_obj_t *self) {
+
```

Comment:
Is it possible to discard the already DMA'd characters? That's what we want to do here.

Reply:
yeah, we could abort transfer, there is an issue with loosing bytes if we do it too fast and often. I will check and test.